### PR TITLE
Add exponential backoff

### PR DESCRIPTION
**Pull request by n-hebert**
This PR as a stage for https://github.com/einaregilsson/beanstalk-deploy/pull/74

### CODE REVIEW

File: beanstalk-deploy.js [hunk 1]
```diff
@@ -129,6 +139,28 @@ function expect(status, result, extraErrorMessage) {
...
+        return await fn();
+    } catch (e) {
+        console.log(`Request ${name} received error: ${e}`);
+
+        // We only keep 400's, and up to a certain effort
+        if (!e.statusCode || !e.statusCode === 400 || depth > MAX_WAIT_DEPTH) {
+            throw e;
+        }
+
+        let backoffLength = 2 ** depth * 10;
```
Comment (evanwatterud) [inline beanstalk-deploy.js]: nit: `const`

File: beanstalk-deploy.js [hunk 2]
```diff
@@ -421,109 +426,80 @@ function waitForDeployment(application, environmentName, versionLabel, start, wa
...
                             }
+                        } else {
+                            if (env.Health === 'Green') {
+                                console.log(`Environment has recovered, health is now ${env.Health}, health status is ${env.HealthStatus}`);
+                                resolve(env);
+                            } else {
+                                if (new Date().getTime() > healThreshold.getTime()) {
+                                    reject(new Error(`Environment still has health ${env.Health} ${waitForRecoverySeconds} seconds after update finished!`));
+                                } else {
+                                    let left = Math.floor((healThreshold.getTime() - new Date().getTime()) / 1000);
```
Comment (evanwatterud) [inline beanstalk-deploy.js]: nit: `const`

File: beanstalk-deploy.js [hunk 3]
```diff
@@ -158,13 +194,13 @@ function deployNewVersion(application, environmentName, versionLabel, versionDes
...
         }
-        expect(404, result); 
+        expect(404, result);
         return uploadFileToS3(bucket, s3Key, fileBuffer);
     }).then(result => {
         expect(200, result);
         console.log(`New build successfully uploaded to S3, bucket=${bucket}, key=${s3Key}`);
-        return createBeanstalkVersion(application, bucket, s3Key, versionLabel, versionDescription);
-    }).then(result => {
+        return callWithRetry("Create Beanstalk version", function () { createBeanstalkVersion(application, bucket, s3Key, versionLabel, versionDescription); }, 0);
```
Comment (evanwatterud) [inline beanstalk-deploy.js]: This is just going to return a Promise, is that intentional?

Edit: whoops missed the promise chaining

**Review by evanwatterud — APPROVED** [review]
LGTM. Could consider converting promise chaining to `async/await` in `deployNewVersion` for thread blocking if you needed that at some point and for consistency, but looks like it doesn't matter now. If you ran a linter there would likely be a lot of things to fix, but again, not sure if that matters here.

File: beanstalk-deploy.js (hunk 2 above)
Comment (n-hebert) [reply beanstalk-deploy.js]: These are just whitespace reformats, so I'll leave as is for an easier merge for [Einar](https://github.com/einaregilsson/beanstalk-deploy).

**Comment by n-hebert** [comment]
> LGTM. Could consider converting promise chaining to `async/await` in `deployNewVersion` for thread blocking if you needed that at some point and for consistency, but looks like it doesn't matter now. If you ran a linter there would likely be a lot of things to fix, but again, not sure if that matters here.

Thanks for the feedback!

I'm new to most of this so I took tips from the net.
We'll see what [Einar](https://github.com/einaregilsson/beanstalk-deploy) has to say about consistency and - yeah - for the linting, same deal. I'm trying to minimize the diff a little.

**Comment by n-hebert** [comment]
I need to replace the date logic for DescribeEvents to stop repeating events, but otherwise we have a pass here: https://github.com/rewindio/foghorn/runs/4526803270?check_suite_focus=true

I don't see any rate limits, which concerns me, so I'll be test driving it a bit more.

**Pull request merged**